### ISSUE TITLE
Investigate torch-spyre Attention Work Optimizations (MHA, non-causal)

### ISSUE THREAD

**Issue by cyang49** [torch-spyre/torch-spyre#2775]
### What happened?

MHA (Multi head attention) benchmark based on `torch.nn.functional.scaled_dot_product_attention` shows large performance gap (>10x) between torch-spyre and sendnn. 

### What did you expect to happen?

Some performance gap is expected, but > 10x suggests that there are misconfigurations than real limitations

### How can we reproduce it?

Run spyre-perf-suite attention benchmark and the newly added fms_granite_micro.mha benchmarks

### Any environmental details we need to know?

First round of tests found:
1. the default query_len=512 and kv_len=4096 creates tensors too large to fit in the 2MB per-core scratchpad. Coarse tiling is required to adjust working set size
2. To emulate the right working set size, we observed that the sendnn processes 4 heads per iteration. Thus, adding a nhead=4 benchmark case can test out the torch-spyre compiler passes.
3. The 4-head problem size decreases the latency on both stacks, but torch-spyre is still about 10x slower

### Relevant log output

```Shell

```

**Comment by cyang49** [comment]
By analyzing the torch-spyre sdsc of the 4-head case, we found several gaps:

1. I added `fms_granite_micro.mha_4h_workdiv_h4_lq8` benchmark case that uses hints to explicitly set work divisions in attention.
2. work-division hints doesn't support when product of split values > 32. I made a new PR #2764 to make it ignore hints when the core budget is used up, instead of raising compile error
3. A restickify on the key tensor due to transposition is observed in torch-spyre but not sendnn
4. A mandatory spill of query tensor after scaling is observed in torch-spyre
5. Different layout dim order for most tensors in torch-spyre compared with sendnn
6. Even though work-division hints are correct, lx planning still fail to pin some tensors. Some of the ops need to be white-listed

**Comment by cyang49** [comment]
For the mha implementation without causal mask, we've resolved most of the issues in work division and lx planning so all intermediate value tensors are pinned in LX. The changes are merged into main. 

@AdnanHoque tested a prototype LX relayout which further brought down the latency by 30% percent.

For performance parity we need to use a torch sdpa implementation that calls the same ops as torch-sendnn

**Comment by cyang49** [comment]
Work division and LX opt are completed through co-optimizer.
LX relayout is tracked in #2963
Closing
